controller/node_controller_test.go

```diff
@@ -2,132 +2,169 @@ package controller
-	KubeNodePressure = "kubeNodePressure"
-)
-
 var (
 	MountPropagationBidirectional = corev1.MountPropagationBidirectional
-)
 
-type NodeTestCase struct {
-	nodes            map[string]*longhorn.Node
-	pods             map[string]*corev1.Pod
-	replicas         []*longhorn.Replica
-	kubeNodes        map[string]*corev1.Node
-	instanceManagers map[string]*longhorn.InstanceManager
+	DefaultOrphanTestNode1 = newOrphan(
+		longhorn.OrphanSpec{
+			NodeID: TestNode1,
+			Type:   longhorn.OrphanTypeReplica,
+			Parameters: map[string]string{
+				longhorn.OrphanDataName: monitor.TestOrphanedReplicaDirectoryName,
+				longhorn.OrphanDiskName: TestDiskID1,
+				longhorn.OrphanDiskUUID: TestDiskID1,
+				longhorn.OrphanDiskPath: TestDefaultDataPath,
+			},
+		},
+		longhorn.OrphanStatus{
+			OwnerID: TestNode1,
+		},
+	)
 
-	expectNodeStatus       map[string]longhorn.NodeStatus
-	expectInstanceManagers map[string]*longhorn.InstanceManager
-	expectOrphans          []*longhorn.Orphan
-}
+	DefaultInstanceManagerTestNode1 = newInstanceManager(
+		TestInstanceManagerName,
+		longhorn.InstanceManagerStateRunning,
+		TestOwnerID1, TestNode1, TestIP1,
+		map[string]longhorn.InstanceProcess{},
+		map[string]longhorn.InstanceProcess{},
+		false,
+	)
+)
 
-func newTestNodeController(lhClient *lhfake.Clientset, kubeClient *fake.Clientset, extensionsClient *apiextensionsfake.Clientset,
-	informerFactories *util.InformerFactories, controllerID string) *NodeController {
-	ds := datastore.NewDataStore(TestNamespace, lhClient, kubeClient, extensionsClient, informerFactories)
+// This data type contains all necessary fixtures for the test, like the mock
+// API clients
+type NodeControllerSuite struct {
+	kubeClient       *fake.Clientset
+	lhClient         *lhfake.Clientset
+	extensionsClient *apiextensionsfake.Clientset
 
-	logger := logrus.StandardLogger()
-	nc := NewNodeController(logger, ds, scheme.Scheme, kubeClient, TestNamespace, controllerID)
-	fakeRecorder := record.NewFakeRecorder(100)
-	nc.eventRecorder = fakeRecorder
-	nc.topologyLabelsChecker = fakeTopologyLabelsChecker
+	informerFactories *util.InformerFactories
 
-	enqueueNodeForMonitor := func(key string) {
-		nc.queue.Add(key)
-	}
-	mon, err := monitor.NewFakeNodeMonitor(nc.logger, nc.ds, controllerID, enqueueNodeForMonitor)
-	if err != nil {
-		return nil
-	}
-	nc.diskMonitor = mon
+	lhNodeIndexer            cache.Indexer
+	lhReplicaIndexer         cache.Indexer
+	lhSettingsIndexer        cache.Indexer
+	lhInstanceManagerIndexer cache.Indexer
+	lhOrphanIndexer          cache.Indexer
 
-	for index := range nc.cacheSyncs {
-		nc.cacheSyncs[index] = alwaysReady
-	}
-	return nc
+	podIndexer  cache.Indexer
+	nodeIndexer cache.Indexer
+
+	controller *NodeController
 }
 
-func fakeTopologyLabelsChecker(kubeClient clientset.Interface, vers string) (bool, error) {
-	return false, nil
+// This data type contains resource that exist in the cluster environment, like
+// nodes and pods
+type TestCase struct {
```

Comment:
Should we name this struct to `NodeTestCase`? I think if we adopt this new test patent for other controllers in this package, the generic `TestCase` naming will be conflicted later on

Reply:
I'll rename it to `NodeControllerFixture` and `NodeControllerExpectation` respectively. If the other test suites also match this structure, we don't need to maintain separate fixture and expectation data structures for each controller, but until I've looked at other tests this makes a lot of sense.